Merged wpt-dbw branches and tl_detector.

khatiba opened the pull request:
I did my best to merge the work done so far for the DBW node and part of the waypoint updater. I had a pretty good implementation of the controller, it has steady throttle and braking, I scanned William's repo and it looks about right. I took most of Xiang's waypoint updater code but left out the traffic light detection for now. This branch just drives the car around at constant velocity and ignores lights.

khatiba committed:
Merged wpt-dbw branches into working controller.

xiangjiang reviewed (commented):
~~This works great on the Highway. However, when I tried it on test lot, it still stuck after about 3/4 of a circle.~~
~~To try it on the test lot, I modified the launch/site.launch to add the simulator bridge (which is present in launch/styx.launch). Then use roslaunch launch/site.launch with the test lot.~~
~~The problem is steering is not enough when the circle is small and angular movement is large.~~
~~Naive solution here is to increase steering no matter what by about 30%.~~
~~Then the car can do infinite loop in test lot.~~

~~Let me know your thoughts.~~

Sorry I think after removing these in the pull request everything works perfectly in both scenarios.
`            # TODO: Temporarily set speed to optimize controller gains`
`            for i in range(len(final_waypoints)):`
`                self.set_waypoint_velocity(final_waypoints, i, 20.0)`

khatiba commented:
Ah, yes, I didn't test the control gains with slower speeds in the test lot, I'll try that out with slower speeds, ~~maybe 25mph or so.~~ 10mph

khatiba commented:
I tested on the church lot and highway but with the speed limits set in the launch files, it performed okay, but I feel like just a small increase in steering would be better. I tried 10,20,30% increases and it was quite a bit better.

khatiba committed:
Remove test speed.

khatiba committed:
Add Xiangs traffic_cb to waypoint updater

khatiba committed:
Optimize check_position.

khatiba committed:
Improve braking while approaching traffic lights.

xiangjiang commented:
Stopping function has very high quality now.
I like the setting overshoot waypoint to target speed -1.0.
There is probably on minor problem I left in the resuming part.
`target_speed = 100/9`
I hardcoded it. I think we can get is from rosparam rather than hardcoding it.
Otherwise, if we include Hari's tl_classifer.py, I think it worth submitting.

xiangjiang commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -21,36 +22,139 @@
+                    new_speed = -1*(current_speed/dist_to_light) * delta + current_speed
+                    self.set_waypoint_velocity(self.all_wps.waypoints, i + self.wp_current_start, new_speed)
+                else:
+                    self.set_waypoint_velocity(self.all_wps.waypoints, j + self.wp_current_start, -1.0)
```
I think there is a typo here, j should be i

khatiba replied:
Good catch! I guess that actually never got called yet. Fixed now.

khatiba commented:
I'll check out the resuming part now.

khatiba committed:
Linearly increase velocity from stop. Add LPF.

khatiba commented:
Xiang, I found some of your commented out code and added my own changes so that the car will linearly increase velocity from a stop, it's basically the reverse of the braking loop. This also respects the speed limit set in ros params. I also forgot about the low pass filter, so I added that to smooth the throttle when accelerating.

xiangjiang commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -29,10 +29,11 @@ class WaypointUpdater(object):
     def __init__(self):
         rospy.init_node('waypoint_updater')
 
+        self.speed_limit = rospy.get_param('/waypoint_loader/velocity', 15) * 0.45 # MPH to m/s
```
I think waypoint loader velocity param is Kilometer per hour.
40 KPH = 100/9 m/s = 24.85 MPH
The multiplier is then 5/18 = 0.277778

khatiba replied:
yes your right, that makes sense.

xiangjiang commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -128,33 +130,35 @@ def traffic_cb(self, msg):
-                    self.set_waypoint_velocity(self.all_wps.waypoints,i+self.wp_current_start+resuming_in_action,target_speed)
+                num_waypoints = len(self.all_wps.waypoints)
+                for j in range(i, num_waypoints - num_resume_wps - 1):
+                    self.set_waypoint_velocity(self.all_wps.waypoints, (j+self.wp_current_start) % num_waypoints, self.speed_limit)
```
C.f. the previous comment, now we don't need to set all the way to the end with speed_limit. Just need to make sure we change previously slowed velocity back.

xiangjiang committed:
tl_detector code with ground truth light state

xiangjiang commented:
I have now included tl_detector.py that I posted in slack. It is not calling get_light_state. 
If Hari is going to add his light classification, please uncomment line 210 and comment line 211, then add your NN with the steps Hari described in slack.

xiangjiang commented on ros/src/waypoint_updater/waypoint_updater.py:
```diff
@@ -21,36 +22,143 @@
+        for i in range(self.last_closest_waypoint, self.last_closest_waypoint + LOOKAHEAD_WPS):# len(self.all_wps.waypoints)):
+            current_distance = dl(pos, self.all_wps.waypoints[i % num_waypoints].pose.pose.position)
+            if current_distance < max_distance:
+                closest_point = i
```
closest_point = i % num_waypoints
Found this out while trying site.launch

khatiba replied:
ah! good catch

khatiba committed:
Fix speed limit conversion and check_point wrap-around.

khatiba committed:
Merge branch 'wpt-dbw' of github.com:khatiba/CarND-System-Integration into wpt-dbw

Commit:
add tl_classifiers both cv_method and carla, using cv_method

Commit:
remove ignored pb file

xiangjiang committed:
add model_detection.pb from referenced source

xiangjiang committed:
ugly fix for errors in ROSBAG replay
